SPT: adds confirm selection button to preview thumbnails

getdave opened the pull request:
This PR addresses https://github.com/Automattic/wp-calypso/issues/36235 and https://github.com/Automattic/wp-calypso/issues/35887 by adding an inner button to each Template thumbnail which appears once the Template has been selected for preview. Clicking on the inner button then triggers the "selection" of that Template and insertion into the Editor.

## Major changes

✅ Individual select & confirm UI interaction pattern for each Template.
✅ Removal of "global" select & confirm UI in favour of individual buttons on a per Template basis.
✅ The normalisation of interaction pattern between "blank" Template and all other Templates.

## Items not addressed here

As this PR forms part of a wider UI update, it's worth noting that the follow closely related issues are **_not_** addressed here

* Any UI update to the visual implementation of Modal itself.
* Change to what is shown for the full-size preview for the "Blank Template".
* Restyling of the individual Template preview thumbnails themselves.

## Important Notes

* The breakpoint where the interaction modal changes from one (mobile) to two (desktop) interactions depends on whether the large preview is visible or not. This itself depends on a CSS media query which makes it laborious to share between CSS and JS. We could parse it out of pseudo-element but this seems like a lot of work for little gain. 

## Testing instructions

* Check out PR 
* Start new Page
* See Template selector
* Click on Template preview (confirm that double / repeated clicking on preview thumbnail itself does not trigger insertion of template)
* See `primary` button appear on top of the Template name
* Click on button to insert Template.
* Confirm the same workflow using only your keyboard.

## Screenshots

#### Desktop

![Screen Capture on 2019-09-24 at 16-26-30](https://user-images.githubusercontent.com/444434/65525961-371df580-dee8-11e9-9cca-e9be8aa178f1.gif)


#### "Mobile" 

This term means "screens which don't have sufficient real estate to accommodate the large preview. 

![Screen Capture on 2019-09-24 at 16-18-52](https://user-images.githubusercontent.com/444434/65525266-2620b480-dee7-11e9-8cf2-734f454cc7b1.gif)

getdave committed:
Adds confirm selection button to preview thumb

getdave committed:
Updates to improve a11y and i18n

Utilise aria and consistent DOM representation between states. Also add translation of previously hard coded strings.

getdave committed:
Removes global Use Template and Cancel buttons

No longer required for Template selection and confirmation because this is now inline with each preview

getdave committed:
Fixes blank selection to respect select and confirm interaction model

Previously “blank” had a unique interaction pattern of 1 click to insert template. Upon Designers feedback switch to same patterns as for all other Templates.

getdave committed:
Update to double or single interaction model for Template selection based on screen size

obenland commented:
Looking good so far @getdave!

I noticed the new button says "Use template" while the mockup shows the template title: "Use About". Should we stay with that?

![Screen Shot 2019-09-24 at 10 54 20 AM](https://user-images.githubusercontent.com/1398304/65537482-26ef2180-deba-11e9-88b7-686c37d5dc10.png)
It would probably be good to account for long button texts ahead of time, possibly by covering more of the preview thumbnail as the lines break?

On small screens I can't get the template to insert after I select it. It gets stuck on "Inserting template…", much like in your gif in the PR description.

obenland commented on apps/full-site-editing/full-site-editing-plugin/starter-page-templates/page-template-modal/components/template-selector-item.js:
```diff
@@ -49,18 +55,44 @@ const TemplateSelectorItem = props => {
+	 * the Template selection UI interaction model should be select and confirm
+	 * or simply a single "tap to confirm". The reason for this is that on larger screens
+	 * the large preview is visible which necessitates the double interaction. Without this
+	 * then only a single tap is necessary.
```
It would be less arbitrary if we differentiated it by whether we show the large preview or not. Which is what it comes down to, no?

getdave replied:
Yeh I agree. It's totally dependent on the presence of the larger preview rather than whether it's "mobile" or desktop. We should bring this into our internal conversations for clarity as well.

Updated code and comments to make this ultra clear.

getdave committed:
Fix not handling template insertion on smaller screens

Addresses https://github.com/Automattic/wp-calypso/pull/36263#issuecomment-534676466

Note that this bug was due to

1. assuming that setState was running sync - we assumed the state update would be ready by the time `handleTemplateConfirmation()` was called
2. handleTemplateConfirmation relies on state for its arguments. By providing explicit arguements we have more control over what we are setting as the template.

getdave committed:
Update comment to make interaction model clearer

getdave commented:
> Looking good so far @getdave!

Thanks 🎉 

> I noticed the new button says "Use template" while the mockup shows the template title: "Use About". Should we stay with that?
> It would probably be good to account for long button texts ahead of time, possibly by covering more of the preview thumbnail as the lines break?

This is intentional and I should have mentioned it.

The issue is the text getting too long and wide. A quick fix was to remove additional text.

I was aiming to avoid this PR ballooning into addressing the wider design changes which should be tackled separately. 

However, give the translation causes this issue to manifest even with my shortened text it now feels legit to address it at some level in this PR. 

> On small screens I can't get the template to insert after I select it. It gets stuck on "Inserting template…", much like in your gif in the PR description.

This was due to us relying on setState being sync when it is - as you know - async. I've made the necessary functions accept explicit args rather than being overly reliant on state. This makes them more flexible for these use cases and avoids over reliance on state.

getdave committed:
Reinstate Template name into confirmation button

getdave committed:
Fix confirm button to wrap to accommodation text

Addresses https://github.com/Automattic/wp-calypso/pull/36263#issuecomment-534676466

retrofox commented on apps/full-site-editing/full-site-editing-plugin/starter-page-templates/page-template-modal/components/template-selector-control.js:
```diff
@@ -28,6 +32,8 @@ export const TemplateSelectorControl = ( {
 	onTemplateSelect = noop,
 	onTemplateFocus = noop,
 	siteInformation = {},
+	currentTemplate = '',
```
minor: side question: what do you think to set up a default `currentTemplate` with `blank`? Just an idea.

retrofox commented on apps/full-site-editing/full-site-editing-plugin/starter-page-templates/page-template-modal/components/template-selector-item.js:
```diff
@@ -49,18 +56,67 @@ const TemplateSelectorItem = props => {
+	 * @param  {string} tplName template name string from the button text
+	 */
+	const handleLabelClick = ( tplSlug, tplName ) => {
+		const largeTplPreviewVisible = window.matchMedia( '(min-width: 660px)' ).matches;
```
I was concerned about this but it seems to be pretty solid: https://caniuse.com/#feat=matchmedia

retrofox commented:
It works really good, just wondering if we could/should handle the whole approach just using CSS instead of js. For instance, hiding the label button, through media queries.
In fact, we could hide it. The gif below shows what happens when the window resizes:
![use-template-button](https://user-images.githubusercontent.com/77539/65602307-0c7c7d00-df7a-11e9-93aa-0615c2f16296.gif)

obenland commented:
Closing in favor of #36323.

getdave commented:
Closing and dropping this approach 😢due to [change in direction](https://github.com/Automattic/wp-calypso/issues/36323)
